File: compiler/nodes/types/placeholder.go
package types

import (
	"wipple/database"
	"wipple/syntax"
	"wipple/visit"
)

type PlaceholderTypeNode struct {
	Facts *database.Facts
}

func (node *PlaceholderTypeNode) GetFacts() *database.Facts {
	return node.Facts
}

func ParsePlaceholderType(parser *syntax.Parser) (*PlaceholderTypeNode, *syntax.Error) {
	span := parser.Spanned()

	_, err := parser.Token("UnderscoreKeyword")
	if err != nil {
		return nil, err
	}

	parser.Commit("in this placeholder type")

	return &PlaceholderTypeNode{
		Facts: database.NewFacts(span()),
	}, nil
}

func (node *PlaceholderTypeNode) Visit(visitor *visit.Visitor) {
	visitType(visitor, node)
}

File: compiler/src/nodes/types/placeholder.rs
use crate::{
    codegen::{Codegen, CodegenCtx, CodegenError},
    database::{Node, NodeRef},
    nodes::visit_type,
    syntax::{ParseError, Parser, TokenKind},
    visit::{Visit, Visitor},
};

#[derive(Debug)]
pub struct PlaceholderTypeNode;

impl Node for PlaceholderTypeNode {}

pub fn parse_placeholder_type(parser: &mut Parser<'_>) -> Result<PlaceholderTypeNode, ParseError> {
    parser.token(TokenKind::UnderscoreKeyword)?;
    Ok(PlaceholderTypeNode)
}

impl Visit for PlaceholderTypeNode {
    fn visit(&self, node: &NodeRef, visitor: &mut Visitor<'_>) {
        visit_type(node, visitor);
    }
}

impl Codegen for PlaceholderTypeNode {
    fn codegen(&self, ctx: &mut CodegenCtx<'_>) -> Result<(), CodegenError> {
        Err(ctx.error())
    }
}
